File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/graphql/schema/GQLRootQueryAdminAccounts.java
package net.nemerosa.ontrack.boot.graphql.schema;

import graphql.schema.DataFetcher;
import graphql.schema.GraphQLFieldDefinition;
import net.nemerosa.ontrack.model.security.Account;
import net.nemerosa.ontrack.model.security.AccountService;
import net.nemerosa.ontrack.model.structure.ID;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static graphql.Scalars.GraphQLInt;
import static graphql.Scalars.GraphQLString;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils.checkArgList;
import static net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils.stdList;
import static org.apache.commons.lang3.StringUtils.contains;

@Component
public class GQLRootQueryAdminAccounts implements GQLRootQuery {

    private final AccountService accountService;
    private final GQLTypeAccount account;

    @Autowired
    public GQLRootQueryAdminAccounts(AccountService accountService, GQLTypeAccount account) {
        this.accountService = accountService;
        this.account = account;
    }

    @Override
    public GraphQLFieldDefinition getFieldDefinition() {
        return newFieldDefinition()
                .name("accounts")
                .type(stdList(account.getType()))
                .argument(
                        newArgument()
                                .name("id")
                                .description("Searching by ID")
                                .type(GraphQLInt)
                                .build()
                )
                .argument(
                        newArgument()
                                .name("name")
                                .description("Searching by looking for a string in the name or the full name")
                                .type(GraphQLString)
                                .build()
                )
                .dataFetcher(adminAccountsFetcher())
                .build();
    }

    private DataFetcher adminAccountsFetcher() {
        return environment -> {
            Integer id = environment.getArgument("id");
            String name = environment.getArgument("name");
            if (id != null) {
                checkArgList(environment, "id");
                return Collections.singletonList(
                        accountService.getAccount(ID.of(id))
                );
            } else {
                Predicate<Account> filter = a -> true;
                // Filter by name
                if (StringUtils.isNotBlank(name)) {
                    filter = filter.and(
                            account -> contains(account.getName(), name) || contains(account.getFullName(), name)
                    );
                }
                // Getting the list
                return accountService.getAccounts().stream()
                        .filter(filter)
                        .collect(Collectors.toList());
            }
        };
    }

}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryAdminAccounts.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars
import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLArgument
import graphql.schema.GraphQLFieldDefinition
import net.nemerosa.ontrack.common.and
import net.nemerosa.ontrack.graphql.support.listType
import net.nemerosa.ontrack.model.security.Account
import net.nemerosa.ontrack.model.security.AccountService
import net.nemerosa.ontrack.model.structure.ID.Companion.of
import org.springframework.stereotype.Component

@Component
class GQLRootQueryAdminAccounts(
        private val accountService: AccountService,
        private val account: GQLTypeAccount,
) : GQLRootQuery {

    override fun getFieldDefinition(): GraphQLFieldDefinition {
        return GraphQLFieldDefinition.newFieldDefinition()
                .name("accounts")
                .type(listType(account.typeRef, nullable = true, nullableItem = false))
                .argument(
                        GraphQLArgument.newArgument()
                                .name(ID_ARGUMENT)
                                .description("Searching by ID")
                                .type(Scalars.GraphQLInt)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(NAME_ARGUMENT)
                                .description("Searching by looking for a string in the name or the full name")
                                .type(Scalars.GraphQLString)
                                .build()
                )
                .argument(
                        GraphQLArgument.newArgument()
                                .name(GROUP_ARGUMENT)
                                .description("Searching by looking for a string in one of the groups the account belongs to")
                                .type(Scalars.GraphQLString)
                                .build()
                )
                .dataFetcher(adminAccountsFetcher())
                .build()
    }

    private fun adminAccountsFetcher(): DataFetcher<*> = DataFetcher { environment: DataFetchingEnvironment ->
        val id = environment.getArgument<Int>(ID_ARGUMENT)
        val name = environment.getArgument<String>(NAME_ARGUMENT)
        val group = environment.getArgument<String>(GROUP_ARGUMENT)
        if (id != null) {
            listOf<Account>(
                    accountService.getAccount(of(id))
            )
        } else {
            var filter: (Account) -> Boolean = { true }
            // Filter by name
            if (!name.isNullOrBlank()) {
                filter = filter and { account ->
                    account.name.contains(name, ignoreCase = true) ||
                            account.fullName.contains(name, ignoreCase = true)

                }
            }
            // Filter by group
            if (!group.isNullOrBlank()) {
                filter = filter and { account ->
                    accountService.getGroupsForAccount(account.id).any { accountGroup ->
                        accountGroup.name.contains(group, ignoreCase = true)
                    }
                }
            }
            // Getting the list
            accountService.accounts.filter(filter)
        }
    }

    companion object {
        const val ID_ARGUMENT = "id"
        const val NAME_ARGUMENT = "name"
        const val GROUP_ARGUMENT = "group"
    }
}
